Add requirement about passing service version (#3809)
Azure/azure-sdk: docs/general/design.md

{% include requirement/MUST id="general-service-apiversion-6" %} ensure that the values of the `ServiceVersion` enumerated value "match" the version strings in the service Swagger definition.

> {% include requirement/MUST id="general-service-apiversion-7" %} replace or add the service version on any URI returned to the service e.g., `Operation-Location`, next page links, etc., with the service version passed to the client. This include both the `api-version` query parameter as well as any versions located within the URI path.

Author: Based on discussions,
```suggestion
{% include requirement/MUST id="general-service-apiversion-7" %} add or replace the `api-version` query parameter on any URI returned by the service e.g., `Operation-Location`, next page links, etc., with the service version passed configured on the client.
```
Does this adequate capture the "MUST"? Should we state that clients "MAY" replace versions in the URL path? IIRC, we didn't explicitly forbid it, but did state it wasn't a requirement. Though, also IIRC, it was discussed that a version in the URL path may indicate an opaque resource URI and should not change.

/cc @m-nash

> 

Reviewer: If we want to make this a "MUST" requirement I think we need to:
a) ensure that our own libraries do this or have a plan to do this, and
b) that we make it easy for service teams that build their own libraries to comply with this guidance.  Which I think means that we need methods in the core SDKs that do the bulk of the work to implement this guideline, and then the only burden on the SDK writers is to make sure to call those methods when appropriate.

Reviewer 2: Changing the api-version query parameter is trivial as a pipeline policy coulde (and perhaps is already) doing this. Having a pipeline policy that fixes the url segment is also trivial (less than 10 lines of code). The pipeline should already be used for retries, telemetry, etc.

Reviewer 3: @JeffreyRichter I think we should go through the SDK arch board to get all language architect sign-off before merging this change to the guidelines.

Reviewer 2: That's fine but this is one of those things that really should happen. It's a bug waiting for a customer to run into it (which may be a while).

Author: I'm getting an SDK archboard meeting set up to discuss this further.

Reviewer 3: Thanks, @heaths!
